core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -153,33 +151,66 @@ private static void add(Map<String, Settings> executorSettings, ExecutorSettings
         executorSettings.put(name, settings);
     }
 
-    private static class ExecutorSettingsBuilder {
-        Map<String, String> settings = new HashMap<>();
+    private static abstract class ExecutorSettingsBuilder<T extends ExecutorSettingsBuilder<T>> {
 
-        public ExecutorSettingsBuilder(String name) {
+        private final Map<String, String> settings;
+
+        protected ExecutorSettingsBuilder(String name, ThreadPoolType threadPoolType) {
+            if (THREAD_POOL_TYPES.get(name) != threadPoolType) {
+                throw new IllegalArgumentException("thread pool [" + name + "] must be of type [" + threadPoolType + "]");
```

Comment:
I'm ok with this. With this we ensure that the proper subclass of ExecutorSettingsBuilder is used for the respective threadpool name. It does not break anything existing.

I am a bit in doubt whether we need ExecutorSettingsBuilder and its subclasses at all. With proper setting validation in the future, we might just get rid of them.

On a totally unrelated note, I noticed that THREAD_POOL_TYPES can be made final.